task_manager/task_manager.hh

```diff
@@ -0,0 +1,364 @@
+
+        void add_child(foreign_task_ptr&& child) {
+            _impl->_children.push_back(std::move(child));
+        }
+
+        void start() {
+            _impl->_status.start_time = db_clock::now();
+            _impl->_status.state = task_manager::task_state::running;
+
+            try {
+                // Background fiber does not capture task ptr, so the task can be unregistered and destroyed independently in the foreground.
+                // After the ttl expires, the task id will be used to unregister the task if that didn't happen in any other way.
+                (void)with_gate(_impl->_module->async_gate(), [f = done(), module = _impl->_module, id = id()] () mutable {
+                    return std::move(f).finally([module, id] {
+                        return sleep_abortable(module->get_task_manager().get_task_ttl(), module->abort_source());
+                    }).then_wrapped([module, id] (auto f) {
+                        f.ignore_ready_future();
+                        module->unregister_task(id);
+                    });
+                });
+                register_task();
+            } catch (...) {
+                finish_failed(std::current_exception());
+            }
+        }
+
+        module_ptr get_module() const noexcept {
+            return _impl->_module;
+        }
+
+        future<progress> get_progress() const {
+            return _impl->get_progress();
+        }
+
+        is_abortable is_abortable() const noexcept {
+            return _impl->is_abortable();
+        };
+
+        future<> abort() noexcept {
+            return _impl->abort();
+        }
+
+        future<> done() const noexcept {
+            return _impl->_done.get_shared_future();
+        }
+
+        void register_task() {
+            _impl->_module->register_task(shared_from_this());
+        }
+
+        void unregister_task() noexcept {
+            _impl->_module->unregister_task(id());
+        }
+
+        void finish() noexcept {
+            if (!_impl->_done.available()) {
+                _impl->_status.end_time = db_clock::now();
+                _impl->_status.state = task_manager::task_state::done;
+                _impl->_done.set_value();
+            }
+        }
+
+        void finish_failed(std::exception_ptr ex, std::string error) noexcept {
+            if (!_impl->_done.available()) {
+                _impl->_status.end_time = db_clock::now();
+                _impl->_status.state = task_manager::task_state::failed;
+                _impl->_status.error = std::move(error);
+                _impl->_done.set_exception(ex);
+            }
+        }
+
+        void finish_failed(std::exception_ptr ex) {
+            finish_failed(ex, fmt::format("{}", ex));
+        }
+    };
+
+    class module : public enable_shared_from_this<module> {
+    protected:
+        task_manager& _tm;
+        std::string _name;
+        task_map _tasks;
+        seastar::gate _gate;
+        int _sequence_number = 0;
```

Comment:
better use `uint64_t`, not that we create new tasks that often, but 32 bits run-out too fast nowadays...